File: frontend/src/types/settings/cleanup-config.ts
/**
 * Settings — Ingestion folder cleanup config domain types.
 *
 * Mirrors backend schemas (schemas/cleanup_config.py):
 *   - CleanupConfigPublic → CleanupConfig
 *   - CleanupConfigUpdate → CleanupConfigInput
 *
 * The cleanup job deletes whole `<base_path>/<dd-mm-yyyy>` folders older than
 * `retention_days` for each ingestion base path (JSW / JVML / Credit Report).
 * Files only — DB ingestion records are never touched.
 */

/** The cleanup policy singleton (no `id` — keyed by "default" on the backend). */
export interface CleanupConfig {
  /** Whether the daily cleanup job is active. */
  enabled: boolean
  /** Date folders strictly older than this many days are deleted (1–365). */
  retention_days: number
  /** Hour of day (0–23, server LOCAL tz) the daily cleanup runs. */
  run_hour: number
  /** ISO-8601 datetime of the last run, or null if it has never run. */
  last_run_at: string | null
  /** Number of folders deleted on the last run. */
  last_deleted_count: number
  /** ISO-8601 datetime of the last save, or null if never configured. */
  updated_at: string | null
}

/** PUT body for saving the cleanup config (server sets the rest). */
export interface CleanupConfigInput {
  enabled: boolean
  retention_days: number
  run_hour: number
}

File: backend/app/models/cleanup_config.py
"""CleanupConfig singleton document (MongoDB collection ``cleanup_configs``, via Beanie).

Drives the daily job that deletes stale ``<base_path>/<dd-mm-yyyy>`` ingestion
folders (files only — DB ingestion records are never touched). The folders it
cleans are the distinct ``base_path``s of the JSW / JVML / Credit Report configs;
this singleton only carries the *policy* (retention window + run hour).
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Annotated

from beanie import Document, Indexed
from pydantic import Field


def _now_utc() -> datetime:
    """Return the current UTC datetime (used as a Field default factory)."""
    return datetime.now(timezone.utc)


class CleanupConfig(Document):
    """Singleton cleanup policy; always fetched/upserted by key="default"."""

    key:                Annotated[str, Indexed(unique=True)] = "default"
    enabled:            bool = False
    retention_days:     int = 5      # delete date folders strictly older than this
    run_hour:           int = 3      # 0–23, LOCAL tz hour the daily job fires
    last_run_at:        datetime | None = None
    last_deleted_count: int = 0
    created_at:         datetime = Field(default_factory=_now_utc)
    updated_at:         datetime = Field(default_factory=_now_utc)

    class Settings:
        name = "cleanup_configs"
